sr5824-msr9732-sm9453/project.md

```diff
@@ -0,0 +1,136 @@
+offer their own proprietary applications, interfaces, APIs and infrastructures, which
+somewhat makes these providers unique. Due to the uniqueness (be it performance or costing)
+companies would want to onboard different set of services onto various providers.
+For example, Salesforce.com, which  started the wave of SaaS back in the late 1900's 
+, has only recently started it's migration from 1st party datacenters to these 3rd party
+providers. Some of their micro-services are being on-boarded to GCP, while a couple of
+others to AWS, resulting in a heterogeneous cloud environment. This heterogeneous 
+environment makes it difficult for companies to change cloud service providers; in 
+any case. Exploring capabilities to support the automated migration from one provider 
+to another is an active, open research area. A lot of people have been pursuing approaches 
+to reduce the impact of vendor lock-in by investigating the cloud migration problem 
+at the level of the VM. However, the migration downtime, decoupling VM from
+underlying systems and security of live channels remain open issues and as a part of
+this project we aim to hypothesize a solution to at least one blatant issue in the
+process of migration.
+
+# Proposed Solution
+
+As we are moving towards heterogeneous cloud, we want to write Cloud-Agnostic
+applications. To make the transition easier, we propose a static code analyser
+to find code snippets specific to a cloud provider using SonarQube Quality
+checks. We chose SonarQube as it is widely being used to check for bugs and
+code issues in large enterprises. Apart from pointing to the position of the change,
+we also provide a correction required to move to a another Cloud provider. 
+For example, let us assume that our code analyser checks AWS-specific 
+code snippet to move to Azure Cloud. Then, we provide the changes required to move to Azure.
+
+The only way to evaluate our service is to check the accuracy of our recommendations.
+We would check the similarity and correctness of our recommended code snippet with the
+actual code snippet provided via the cloud service provider (it's no rocket science, to
+google a certain part of code).
+Example: If we have 4 recommendations, out of which 3 are correct, we can calculate our
+accuracy and generate a confusion matrix based on these statistics.
+
+To better explain our hypothesis, let us consider these examples:
+
+As in the case of **batch jobs**, a capability provided by all public clouds, or at least the ones that are
+pertinent to the scope of this project.
+
+Usually, batch jobs require spinning up of cluster of workers, followed by jobs submitted to
+these workers. When we look at the implementations of this API across all the cloud platforms,
+the underlying principle is the same. 
+
+Let's consider a particular case, integration of batch job APIs with Java. Each cloud provider has it's
```

Comment:
Q. Will you be extending this tool to work for code written in other languages as well, or just Java?

Reply:
To start off with, we are targetting Java. The reason being the complexity as you mentioned of identifying different dependencies in different languages. However, once we have it set it up for one language, extending the concept to another language won't be a big overhead. The underlying concept is the same, the semantics and the syntax would definitely differ. We are aiming on developing an ideology and having a proof of concept that would help us further explore this particular topic.